### PR TITLE
Weakpoints as objects

### PR DESCRIPTION
**Pull request by I-am-Erk**
#### Summary

Content "Now you can stab a zombie in the eye and it will go blind"

#### Purpose of change

With #51770, #51961, and #51979, @Joshua-Chin has been implementing a robust framework for detailed monster weakpoints and effects from hitting them, and I am way behind on actually using them in monsters!

This turned out to be a [very, very bulky system](https://github.com/CleverRaven/Cataclysm-DDA/pull/52084) to implement inline with the detail I wanted, so I [begged for someone to help fix that](https://github.com/CleverRaven/Cataclysm-DDA/issues/52894) and now, [dseguin has come to my aid](https://github.com/CleverRaven/Cataclysm-DDA/pull/53452).

#### Describe the solution
This PR will add new, much more detailed weakpoint information to all humanoid zombies and most other humanoid monsters. I am not looking at non-humanoid monsters currently, because at least with zombies I can carry the same weakpoints forward to all the monsters.

* Hitting the legs has a chance of knockodwn
* Hitting the arms has a chance to stagger your opponent (momentary drop in speed)
* Hitting the head has a chance of stun
* Hitting the eyes has a chance of blinding, and if you do enough damage it can be permanent

In general, most weakpoints are harder to hit at range than in melee. If they are off the front-facing part of the zombie (eg an armpit) they are much harder. Small weakpoints are harder to hit with cut/bash weapons, and easier with stab/ranged attacks. Large ones like legs are easier to hit with wide sweeping cut/blunt attacks. Weakpoints on the extremities have decreased critical hit damage.

Also: if you are being grabbed by a humanoid, you have a higher chance of hitting the arms of the grabber, and therefore a higher chance of causing the Staggered effect. This effect now cancels grabs. In other words, you can struggle with the arms of your opponent and potentially dislodge them.

#### Describe alternatives you've considered
There are some possible weakpoints I have ignored for now, like hitting a feral in the chest and leaving it winded, or stabbing a zombie in the throat. The modular setup of these makes it easy to consider adding more later.

The actual balance of these still needs testing. In general I suspect I am lowballing the difficulty, we'll have to see.

#### Testing
Built, ran, and tested. All features appear to be functional. The actual balance will require more testing

#### Additional context
This is so awesome.

We may wish to increase the base armour on some of these zombies, but I would suggest first to do some playtesting to see how much trickier I have made their weakpoints. The effective DPS tests have, for the time being, been set to ignore weakpoints altogether. We should develop new tests for weakpoints, and also buff armour as the weakpoints tests now find effective dps way higher against many zombies.

### CODE REVIEW

**Comment by dseguin** [comment]
Something important to note:
The JSON for weakpoint sets must load before the JSON for monsters because of the `"weakpoint_sets"` dependency. To do that, just be sure that the name of the file/directory containing the weakpoint definitions is lexicographically before the monster definitions. You can also get around that by putting the weakpoints in `data/json/mon_weakpoints/` ("mon_weakpoints" comes before "monsters").

It's kind of unintuitive but the loading order works in mysterious ways.

**Comment by I-am-Erk** [comment]
Thanks! Let me know if you think there is any function in your new implementation that I should be using here but am not.

**Comment by I-am-Erk** [comment]
Needs some linting and testing but the content addition is good for now I think. If a reviewer can spot any humanoids I've missed, I'd appreciate it.

File: data/json/monster_weakpoints/generic_weakpoints.json [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        "coverage": 5,
+        "difficulty": { "melee": 4, "ranged": 5 },
+        "coverage_mult": { "broad": 0.5 },
+        "armor_mult": { "all": 0.25 } 
+      },
+      { "name": "a soft fold in the thick hide", 
+        "coverage": 10,
+        "difficulty": { "melee": 3, "ranged": 5 },
+        "coverage_mult": { "point": 0.75 },
+        "armor_mult": { "all": 0.5 }
```
Comment (dseguin) [inline data/json/monster_weakpoints/generic_weakpoints.json:16]:
```suggestion
      {
        "name": "a small gap in the thick hide",
        "coverage": 5,
        "difficulty": { "melee": 4, "ranged": 5 },
        "coverage_mult": { "broad": 0.5 },
        "armor_mult": { "all": 0.25 }
      },
      {
        "name": "a soft fold in the thick hide",
        "coverage": 10,
        "difficulty": { "melee": 3, "ranged": 5 },
        "coverage_mult": { "point": 0.75 },
        "armor_mult": { "all": 0.5 }
```

File: data/json/monster_weakpoints/generic_weakpoints.json [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        "coverage": 5,
+        "difficulty": { "melee": 4, "ranged": 5 },
+        "coverage_mult": { "broad": 0.5 },
+        "armor_mult": { "all": 0.25 } 
+      },
+      { "name": "a joint in the bony plates", 
+        "coverage": 10,
+        "difficulty": { "melee": 3, "ranged": 7 },
+        "coverage_mult": { "ranged": 0.25 },
+        "armor_mult": { "all": 0.5 }
```
Comment (dseguin) [inline data/json/monster_weakpoints/generic_weakpoints.json:34]:
```suggestion
      {
        "name": "a small gap in the bony plates",
        "coverage": 5,
        "difficulty": { "melee": 4, "ranged": 5 },
        "coverage_mult": { "broad": 0.5 },
        "armor_mult": { "all": 0.25 }
      },
      {
        "name": "a joint in the bony plates",
        "coverage": 10,
        "difficulty": { "melee": 3, "ranged": 7 },
        "coverage_mult": { "ranged": 0.25 },
        "armor_mult": { "all": 0.5 }
```

File: data/json/monster_weakpoints/generic_weakpoints.json [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            "effect": "blind",
+            "permanent": true,
+            "chance": 20,
+            "message": "The %s's eyes are obliterated!",
+            "damage_required": [ 41, 100 ]
+          }
+        ]
+      }
+    ]
+  },
```
Comment (dseguin) [inline data/json/monster_weakpoints/generic_weakpoints.json:63]:
```suggestion
  }
```

File: data/json/monster_weakpoints/humanoid_weakpoints.json [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+  },
+  {
+    "type": "weakpoint_set",
+    "id": "wps_humanoid_turnout_gear",
+    "weakpoints": [
+      {
+        "id": "arm",
+		"name": "a tear in its turnout gear",
+		"armor_mult": { "physical": 0.5, "acid": 0.5 }, 
+        "difficulty": { "ranged": 3, "melee": 1 },
```
Comment (dseguin) [inline data/json/monster_weakpoints/humanoid_weakpoints.json:225]:
```suggestion
      {
        "id": "arm",
        "name": "a tear in its turnout gear",
        "armor_mult": { "physical": 0.5, "acid": 0.5 },
        "difficulty": { "ranged": 3, "melee": 1 },
```

File: data/json/monster_weakpoints/humanoid_weakpoints.json [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+            "damage_required": [ 31, 100 ]
+          }
+        ],
+		"coverage": 5 
+	  },
+      {
+        "id": "leg",
+        "name": "a tear in its turnout gear",
+		"armor_mult": { "physical": 0.5, "acid": 0.5 }, 
+        "difficulty": { "ranged": 3, "melee": 1 },
```
Comment (dseguin) [inline data/json/monster_weakpoints/humanoid_weakpoints.json:242]:
```suggestion
        ],
        "coverage": 5
      },
      {
        "id": "leg",
        "name": "a tear in its turnout gear",
        "armor_mult": { "physical": 0.5, "acid": 0.5 },
        "difficulty": { "ranged": 3, "melee": 1 },
```

File: data/json/monster_weakpoints/humanoid_weakpoints.json [hunk 6]
```diff
@@ -0,0 +1,551 @@
...
+      {
+        "id": "eye",
+        "name": "the goggles",
+        "armor_mult": { "physical": 1 },
+        "coverage": 1,
+        "coverage_mult": { "ranged": 0.5 },
+        "difficulty": { "point": 7, "broad": 9 },
+        "effects": [
+          { "effect": "blind", "duration": [ 1, 2 ], "chance": 1, "message": "The %s is blinded!  The goggles, they did nothing!", "damage_required": [ 20, 100 ] }
+        ],
```
Comment (dseguin) [inline data/json/monster_weakpoints/humanoid_weakpoints.json:493]:
```suggestion
        "effects": [
          {
            "effect": "blind",
            "duration": [ 1, 2 ],
            "chance": 1,
            "message": "The %s is blinded!  The goggles, they did nothing!",
            "damage_required": [ 20, 100 ]
          }
        ],
```

File: data/json/monster_weakpoints/humanoid_weakpoints.json [hunk 7]
```diff
@@ -0,0 +1,551 @@
...
+      {
+        "id": "eye",
+        "name": "the goggles",
+        "armor_mult": { "physical": 1 },
+        "coverage": 1,
+        "coverage_mult": { "ranged": 0.5 },
+        "difficulty": { "point": 7, "broad": 9 },
+        "effects": [
+          { "effect": "blind", "duration": [ 1, 2 ], "chance": 1, "message": "The %s is blinded!  The goggles, they did nothing!", "damage_required": [ 20, 100 ] }
+        ]
```
Comment (dseguin) [inline data/json/monster_weakpoints/humanoid_weakpoints.json:511]:
```suggestion
        "effects": [
          {
            "effect": "blind",
            "duration": [ 1, 2 ],
            "chance": 1,
            "message": "The %s is blinded!  The goggles, they did nothing!",
            "damage_required": [ 20, 100 ]
          }
        ]
```

File: data/json/monsters/zed-winged.json [hunk 8]
```diff
@@ -22,6 +22,18 @@
...
+      {
+        "id": "arm",
+        "name": "the wing",
+        "crit_mult": { "all": 0.75 },
+        "difficulty": { "ranged": 2, "melee": 1 },
+        "coverage_mult": { "point": 0.25 },
+        "effects": [ { "effect": "staggered", "chance": 15, "message": "The %s is knocked off-balance!", "damage_required": [ 10, 100 ] } ],
+        "coverage": 12
+      }
+    ]
```
Comment (dseguin) [inline data/json/monsters/zed-winged.json:36]:
```suggestion
    ],
```

File: data/json/monsters/zed-winged.json [hunk 9]
```diff
@@ -75,6 +87,18 @@
...
```
Comment (dseguin) [inline data/json/monsters/zed-winged.json:101]:
```suggestion
    ],
```

File: data/json/monsters/zed_soldiers.json [hunk 10]
```diff
@@ -648,8 +411,15 @@
...
+    "weakpoint_sets": [ "wps_humanoid_full_helmet" ],
     "weakpoints": [
-      { "name": "a gap in the armor", "armor_mult": { "bullet": 0.5, "cut": 0.5, "stab": 0.25, "bash": 0.25 }, "coverage": 4 }
+      {
+       "name": "a crack in the armor",
+       "armor_mult": { "bullet": 0.5, "cut": 0.5, "stab": 0.25, "bash": 0.25 },
+       "coverage": 4,
+       "coverage_mult": { "broad": 0.75 },
+       "difficulty": { "ranged": 7, "melee": 5 }
+      }
```
Comment (dseguin) [inline data/json/monsters/zed_soldiers.json:422]:
```suggestion
      {
        "name": "a crack in the armor",
        "armor_mult": { "bullet": 0.5, "cut": 0.5, "stab": 0.25, "bash": 0.25 },
        "coverage": 4,
        "coverage_mult": { "broad": 0.75 },
        "difficulty": { "ranged": 7, "melee": 5 }
      }
```

**Review by dseguin — COMMENTED** [review]
Just needs some linting and syntax fixes.

**Pull request merged**